---
title: "Collider Bias in GWAS"
subtitle: "STAT 494: Statistical Genetics"
author: "Charles Batsaikhan, Mikaely Evans, and Julia Ross"
date: today
format:
  html:
    toc: true
    toc-depth: 3
    embed-resources: true
---

```{r load-packages}
#| message: false
library(snpStats)
library(tidyverse)
library(broom)
library(qqman)
```

# Data

```{r file-paths}
# update file paths
bed <- 'data/1_QC_GWAS/HapMap_3_r3_1.bed'
bim <- 'data/1_QC_GWAS/HapMap_3_r3_1.bim'
fam <- 'data/1_QC_GWAS/HapMap_3_r3_1.fam'
```

```{r read-data}
hapmap <- read.plink(bed, bim, fam)
```

```{r explore-data}
class(hapmap)
names(hapmap)
```

```{r}
X <- as(hapmap$genotypes, "numeric")

maf <- col.summary(hapmap$genotypes)$MAF

X.clean <- X[, maf>0]

```


\
\
\


# Simulation

### Selecting a single SNP to use as a our exposure

```{r}
set.seed(111) #for reproducability

#Randomly choosing 1 SNP from the cleaned genotype 
snp_index <- sample(ncol(X.clean), 1)
snp_exposure <- X.clean[, snp_index]
snp_name <- colnames(X.clean)[snp_index]
cat("Using SNP:", snp_name)

```


### Generating Sex
```{r}
n <- nrow(X.clean)
sex_sim <- rbinom(n, 1, 0.5) #independet of genotype, basically 50/50 chance
```


### Generating Height as the Collider

```{r}

#assuming a baseline of 5 foot 4 for females and 5 foot 9 for males
#making a SNP have an effect (2 per copy) on height
#adding some error (random residual)

height_sim <- 64 + 5 * sex_sim + 2 * snp_exposure + rnorm(n, mean = 0, sd = 3)
```


### Simulating Data

```{r}
iterations <- 500

pvalues_unadj <- numeric(iterations)
pvalues_adj <- numeric(iterations)

n <- nrow(X.clean)

set.seed(200)

for (i in 1:iterations) {
  sex_sim <- rbinom(n,size=1, prob=0.5)
  height_sim <- 64 + 5 * sex_sim + 2 * snp_exposure + rnorm(n, mean = 0, sd = 3)
  
  sim_data <- data.frame(
    SNP = snp_exposure,
    Sex = sex_sim,
    Height = height_sim
  )
  
  # Sex~SNP
  mod_unadj <- glm(Sex ~ SNP, data = sim_data, family = binomial)
  pvalues_unadj[i] <- summary(mod_unadj)$coefficients["SNP", "Pr(>|z|)"]
  
  # Sex~SNP+HEIGHT
  mod_adj <- glm(Sex ~ SNP + Height, data = sim_data, family = binomial)
  pvalues_adj[i] <- summary(mod_adj)$coefficients["SNP", "Pr(>|z|)"]
  
  
}
```


\
\
\


# Plots

## QQ Plots

```{r}
library(qqman)


qq(pvalues_unadj, main = "Sex~SNP QQ Plot")


qq(pvalues_adj, main = "Sex~SNP+HEIGHT QQ Plot")
```





















